Author: Guilherme G. Piccoli <gpiccoli@canonical.com>
Bug-Ubuntu: https://launchpad.net/bugs/1750013
Forwarded: not-needed
Reviewed-By: Guilherme G. Piccoli <gpiccoli@canonical.com>
Last-Update: 2018-02-16
Description: logind: fix session not added to gc and other memleaks
  1) Proactively add session to gc on its Release event in order to prevent
  entire session's leaks in logind due to not adding closing sessions to gc;

  2) Fix small leak in session_free(), by freeing the reset_controllers ptr;

  3) Fix small leak in cgmanager glue code.

--- a/src/login/logind-session.c
+++ b/src/login/logind-session.c
@@ -93,6 +93,7 @@
 
         free(s->cgroup_path);
         strv_free(s->controllers);
+        strv_free(s->reset_controllers);
 
         free(s->tty);
         free(s->display);
--- a/src/login/logind.c
+++ b/src/login/logind.c
@@ -1503,6 +1503,8 @@
                 if (session_check_gc(session, drop_not_started) == 0) {
                         session_stop(session);
                         session_free(session);
+                } else if (session_get_state(session) == SESSION_CLOSING) {
+                        session_add_to_gc_queue(session);
                 }
         }
 
--- a/src/shared/cgroup-util.c
+++ b/src/shared/cgroup-util.c
@@ -44,7 +44,7 @@
         _cleanup_free_ char *fs = NULL;
         FILE *f;
         int r;
-        char *value = NULL;
+        _cleanup_free_ char *value = NULL;
         char *template = NULL;
         int fd;
 
@@ -111,7 +111,7 @@
         _cleanup_free_ char *fs = NULL;
         FILE *f;
         int r;
-        char *value = NULL;
+        _cleanup_free_ char *value = NULL;
         char *template = NULL;
         int fd;
 
--- a/src/login/logind-dbus.c
+++ b/src/login/logind-dbus.c
@@ -1781,6 +1781,7 @@
                 this call explicitly turn off the FIFO logic, so that
                 the PAM code can finish clean up on its own */
                 session_remove_fifo(session);
+                session_add_to_gc_queue(session);
 
                 reply = dbus_message_new_method_return(message);
                 if (!reply)
